torch/csrc/jit/codegen/cuda/lower_vectorize_welford.cpp

```diff
@@ -0,0 +1,567 @@
+          UnaryOpType::Set,
+          reciprocal,
+          GpuLower::current()->kernel()->zeroVal()));
+
+      // if (pred) reciprocal = 1 / new_count_float;
+      auto reciprocal_ite = IrBuilder::create<kir::IfThenElse>(
+          IrBuilder::create<kir::Predicate>(pred));
+      registerInsertBeforeInnerMostLoop(reciprocal_ite);
+      registerInsertBefore(
+          nullptr, reciprocal_expr, &(reciprocal_ite->thenBody()));
+    }
+
+    auto vectorized_wop = IrBuilder::create<kir::VectorizedWelfordOp>(
+        wop->outputTriplet(),
+        wop->inputTriplet(),
+        wop->initTriplet(),
+        new_count,
+        reciprocal,
+        pred);
+
+    return vectorized_wop;
+  }
+
+  // Declare a scalar variable of type dt and insert its allocation
+  Val* defineScalar(DataType dt) {
+    Val* val = IrBuilder::newScalar(dt);
+
+    auto alloc = IrBuilder::create<kir::Allocate>(
+        val, MemoryType::Local, GpuLower::current()->kernel()->oneVal());
+
+    registerInsertBeforeInnerMostLoop(alloc);
+
+    return val;
+  }
+
+  // Hoist the count TensorIndex out of the innermost loop, assuming
+  // it is invariant within the loop. The loop index can be replaced
+  // with what ever value within the loop range since it is
+  // independent of the loop index.
+  kir::TensorIndex* hoistCount(kir::TensorIndex* out_N) {
+    TORCH_INTERNAL_ASSERT(!for_loops_.empty());
+    auto innermost_loop = for_loops_.back();
+    const auto& original_indices = out_N->indices();
+    std::unordered_map<Val*, Val*> index_replacement_map;
+    index_replacement_map.emplace(
+        innermost_loop->index(), GpuLower::current()->kernel()->zeroVal());
+
+    std::vector<Val*> indices_zero;
+    std::transform(
+        original_indices.begin(),
+        original_indices.end(),
+        std::back_inserter(indices_zero),
+        [&](Val* original_index) {
+          return ir_utils::replaceValInIndexVal(
+              original_index, index_replacement_map);
+        });
+
+    auto hoisted_count =
+        IrBuilder::create<kir::TensorIndex>(out_N->view(), indices_zero);
+
+    return hoisted_count;
+  }
+
+  // Check if all exprs in the same scope as a WelfordOp have the same
+  // predicate, thus allowing the predicates of all of the exprs can
+  // be hoisted collectively. The motivation of this analysis is to
+  // transform exprs from:
+  //
+  // for () {
+  //   if (pred1) {
+  //     expr1;
+  //   }
+  //   if (pred2) {
+  //     welfordOp;
+  //   }
+  // }
+  //
+  // to below when pred1 == pred2.
+  //
+  // if (pred1) {
+  //   for () {
+  //     expr1;
+  //     welfordOp;
+  //   }
+  // }
+  //
+  // Rather than:
+  //
+  // bool pred2 = ...;
+  // for () {
+  //   if (pred1) {
+  //     expr1;
+  //   }
+  //   welfordOp(pred2, ...);
+  // }
+  //
+  // The pattern matching here only works with a simple sequence of
+  // expressions but should be sufficient as it is likely the most
+  // common situation.
```

Comment:
Seems like this could be a nice generic pass like (a future) index hoisting. Seems to me it's a cleanup of unrolling loops, i.e. trying to simplify the else of the unrolled ite. Noting as future work.

Reply:
Agreed. This particular implementation is really simplistic as I don't expect nothing complex loop body should appear with a serial WelfordOp, so we would need to consider a wider variation of patterns.